pippo-core/src/main/java/ro/pippo/core/AbstractTemplateEngine.java

```diff
@@ -0,0 +1,173 @@
+import java.util.Map;
+
+/**
+ * Convenience abstract implementation of {@link TemplateEngine}
+ *
+ * A configuration configuration target <code>CT</code> is a template engine specific component
+ * that needs to be setup for the engine to function correctly.
+ *
+ * To use the convenience methods in this class, implementations must override {@link #init(Application)}
+ * and call this class' implementation before performing any of their initialization.
+ *
+ * @param <CT> configuration target for custom initialization, implementations could use this
+ *
+ * @see TemplateEngine
+ *
+ * @author Ranganath Kini
+ */
+public abstract class AbstractTemplateEngine<CT> implements TemplateEngine {
+
+    private Languages languages;
+    private Messages messages;
+    private Router router;
+    private PippoSettings pippoSettings;
+
+    private String fileExtension;
+    private String templatePathPrefix;
+
+    /**
+     * Performs common initialization for template engines
+     *
+     * Implementations must override this method to do their own template engine specific initialization. To
+     * use the convenience of this class, implementations must invoke this class's implementation before
+     * performing their own initialization.
+     *
+     * @param application reference to the Pippo {@link Application} that can be used to retrieve settings
+     *                     and other settings for the initialization
+     */
+    @Override
+    public void init(Application application) {
+        languages = application.getLanguages();
+        messages = application.getMessages();
+        router = application.getRouter();
+        pippoSettings = application.getPippoSettings();
+
+        fileExtension = pippoSettings.getString(PippoConstants.SETTING_TEMPLATE_EXTENSION, getDefaultFileExtension());
+        templatePathPrefix = pippoSettings.getString(PippoConstants.SETTING_TEMPLATE_PATH_PREFIX, TemplateEngine.DEFAULT_PATH_PREFIX);
+    }
+
+    /**
+     * Performs additional initialization on the specified template engine specific configuration target
+     *
+     * This method serves as a hook for implementation to support additional fine-grained extension to initialize
+     * a template engine specific configuration object.
+     *
+     * Implementations can support this by invoking this method from {@link #init(Application)}.
+     * Supporting this is completely optional.
+     *
+     * The default implementation throws a {@link NotImplementedException}.
+     *
+     * @param application reference to the Pippo {@link Application} the can be used to access configuration details
+     * @param configurationTarget reference to the template engine specific configuration object that needs initialization
+     *
+     * @throws NotImplementedException
+     *          if no additional initialization is supported by the implementation
+     */
+    protected void init(Application application, CT configurationTarget) {
```

Comment:
I don't need this method. In my concrete implementation (`PebbleTemplateEngine` for example) I prefer below approach:
```java
public void init(Application application) {
    super.init(application);

    // specific code ...

    // allow custom initialization
    init(application, builder);
}
```

Reply:
Sure makes sense. So if a specific implementation wants to support further extension to the initialization logic, let them provide their own extension point. This will also help get rid of the generic type.

Reply:
Yes. You read my mind.